---
title: "Data analysis first dataset - GEX, 26.10.2023"
author:
  - "trnguyen@ukaachen.de"
date: "Last update on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    keep_md: true
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 3
    number_sections: false
    theme: lumen
---

```{css zoom-lib-src, echo = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```

```{js zoom-jquery, echo = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

<style type="text/css">
    div.datatables { height: auto !important;}
</style>


```{r echo=FALSE, warning=FALSE, results='hide', message=FALSE}
#####----------------------------------------------------------------------#####

# clean up 
gc()
rm(list = ls())

source("/media/hieunguyen/HNSD01/src/sc_bulk_BCR_data_analysis/VDJ_helper_functions.R")
# install.packages("https://cran.r-project.org/src/contrib/cpp11_0.5.0.tar.gz", type = "source", repos = NULL)
# install.packages("https://cran.r-project.org/src/contrib/igraph_2.1.1.tar.gz", type = "source", repos = NULL)

#####----------------------------------------------------------------------#####
# CONFIGURATIONS AND PREPARATION
#####----------------------------------------------------------------------#####
analysis.case <- "1st_dataset_removed_7_9_removed_16_removed_9.with_reInt.res1"
thres <- 0.85

source("/media/hieunguyen/HNSD01/src/bcr_data_analysis/1st_2nd_datasets/00_helper_functions.R")
scrna_pipeline_src <- "/media/hieunguyen/HNSD01/src/src_pipeline/scRNA_GEX_pipeline/processes_src"
source(file.path(scrna_pipeline_src, "import_libraries.R"))
source(file.path(scrna_pipeline_src, "helper_functions.R"))
source(file.path(scrna_pipeline_src, "s8_integration_and_clustering.R"))

#####----------------------------------------------------------------------#####
# CONFIGURATIONS AND PREPRATIONS
#####----------------------------------------------------------------------#####
outdir <- "/media/hieunguyen/HNSD_mini/outdir/sc_bulk_BCR_data_analysis_v0.1"
PROJECT <- "1st_2nd_BSimons_Datasets"

path.to.output <- file.path(outdir, PROJECT, "data_analysis", "NEW")

path.to.01.output <- file.path(path.to.output, "01_output")
dir.create(path.to.01.output, showWarnings = FALSE, recursive = TRUE)

#### list of output paths
path.to.clone.distribution <- file.path(path.to.01.output, "clone_distribution_UMAP")
dir.create(path.to.clone.distribution, showWarnings = FALSE, recursive = TRUE)

s.obj <- readRDS(file.path(outdir, 
                           PROJECT, 
                           "data_analysis", 
                           "01_output",
                           "1st_dataset_removed_7_9_removed_16_removed_9.with_reInt.res1", 
                           "1st_dataset_removed_7_9_removed_16_removed_9.with_reInt.res1.rds"))

if ("svglite" %in% installed.packages() == FALSE){
  install.packages("svglite")
}

##### read the summary clone information with mutation rate
clonedf <- read.csv(file.path(outdir, "VDJ_output", "04_output", "full_clonedf_with_mutation_rate.csv"))
clonedf <- subset(clonedf, clonedf$dataset.name == "1st_2nd_BSimons_Datasets")
clonedf <- subset(clonedf, clonedf$id %in% unique(s.obj$name))
clonedf <- clonedf %>% rowwise() %>%
  mutate(barcode = sprintf("%s_%s_%s", id, id, barcode))

#####----------------------------------------------------------------------#####
##### define clones
#####----------------------------------------------------------------------#####
#> although this is a single cell dataset, we define the clones using the same
#> definition we used in bulk dataset
#> Update the package igraph to version 2.1.1 in the docker image trex
#> since the old version has some problem while running assign_clusters_to_sequences
clonesets <- clonedf
new.clonesets <- data.frame()
for (input.VJ.combi in unique(clonesets$VJ.len.combi)){
  tmpdf <- subset(clonesets, clonesets$VJ.len.combi == input.VJ.combi)
  seqs <- unique(tmpdf$aaSeqCDR3)
  if (length(seqs) >= 2){
    cluster.output <- assign_clusters_to_sequences(seqs, threshold = thres)$res
    tmpdf[[sprintf("VJcombi_CDR3_%s", thres)]] <- unlist(lapply(
      tmpdf$aaSeqCDR3, function(x){
        return(sprintf("%s_%s", input.VJ.combi, subset(cluster.output, cluster.output$seq == x)$cluster))
      }
    ))    
  } else {
    tmpdf[[sprintf("VJcombi_CDR3_%s", thres)]] <- sprintf("%s_1", input.VJ.combi)
  }
  new.clonesets <- rbind(new.clonesets, tmpdf)
}

row.names(new.clonesets) <- NULL
new.clonesets <- new.clonesets[!duplicated(new.clonesets$barcode), ]
new.clonesets <- new.clonesets %>% column_to_rownames("barcode")
new.clonesets <- new.clonesets[row.names(s.obj@meta.data),] 
s.obj <- AddMetaData(object = s.obj, col.name = sprintf("VJcombi_CDR3_%s", thres), metadata = new.clonesets[[sprintf("VJcombi_CDR3_%s", thres)]])

all.cells <- s.obj@meta.data[[sprintf("VJcombi_CDR3_%s", thres)]]
names(all.cells) <- row.names(s.obj@meta.data)

with.clone.cells <- all.cells[is.na(all.cells) == FALSE] %>% names()
without.clone.cells <- all.cells[is.na(all.cells) == TRUE] %>% names()

#####----------------------------------------------------------------------#####
##### Get reference genes
#####----------------------------------------------------------------------#####
ref.gene.config <- "/media/hieunguyen/HNSD01/src/sc_bulk_BCR_data_analysis/ref_gene_config.R"
source(ref.gene.config) # path to the configuration file for the input BCR reference genes
  cellranger.ref <- readDNAStringSet(ref.genes$`10x`)
  names(cellranger.ref) <- to_vec(
    for (item in names(cellranger.ref)) str_split(str_split(item, " ")[[1]][[1]], "[|]")[[1]][[2]])
```

# UMAP

## All samples, grouped by clusters
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
DimPlot(object = s.obj, reduction = "INTE_UMAP", label = TRUE, label.box = TRUE, repel = TRUE)
```

## All samples, highlighted cells with clone VDJ information
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
DimPlot(object = s.obj, reduction = "INTE_UMAP", label = TRUE, label.box = TRUE, repel = TRUE,
        cols.highlight = c("red", "gray"), cells.highlight = list("with.Clone" = with.clone.cells, "without.Clone" = without.clone.cells))
```

# VDJ information
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
sc.clonedf <- table(s.obj@meta.data[[sprintf("VJcombi_CDR3_%s", thres)]]) %>% as.data.frame()
colnames(sc.clonedf) <- c("Clone", "count")
sc.clonedf <- sc.clonedf %>% arrange(desc(count))
sc.clonedf$mutation_rate <- unlist(lapply(
  sc.clonedf$Clone, function(x)
  {
    v.gene <- str_split(x, "_")[[1]][[1]]
    j.gene <- str_split(x, "_")[[1]][[2]]
    cdr3.len <- str_split(x, "_")[[1]][[3]] %>% as.numeric()
    v.gene.len <- cellranger.ref[[v.gene]] %>% nchar()
    j.gene.len <- cellranger.ref[[j.gene]] %>% nchar()
    total.len <- v.gene.len + cdr3.len + j.gene.len
    tmpdf <- subset(new.clonesets, new.clonesets[[sprintf("VJcombi_CDR3_%s", thres)]] == x)$num_mutation %>% as.numeric()
    return(mean(tmpdf)/total.len)
  }
))

##### add the mutation rate information to seurat object
meta.data <- s.obj@meta.data %>% rownames_to_column("barcode")
meta.data$mutation_rate <- unlist(lapply(
  meta.data[[sprintf("VJcombi_CDR3_%s", thres)]], function(x){
    if (is.na(x) == FALSE){
      return(subset(sc.clonedf, sc.clonedf$Clone == x)$mutation_rate)      
    } else {
      return(NA)
    }
  }
))
meta.data <- meta.data %>% column_to_rownames("barcode")
meta.data <- meta.data[row.names(s.obj@meta.data), ]
s.obj <- AddMetaData(object = s.obj, col.name = "mutation_rate", metadata = meta.data$mutation_rate)
```

## Clone size vs number of mutations
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
sc.clonedf %>% ggplot(aes(x = "All clones", y = mutation_rate)) + geom_boxplot() + theme_pubr()
```

## Mutation rate on UMAP
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
library(viridis)
# we remove cells without mutation rate before plotting
with.mutation.rate.cells <- subset(s.obj@meta.data, is.na(s.obj@meta.data$mutation_rate) == FALSE) %>% row.names()
with.mutation.rate.s.obj <- subset(s.obj, cells = with.mutation.rate.cells)
FeaturePlot(object = with.mutation.rate.s.obj, reduction = "INTE_UMAP", features = c("mutation_rate"), pt.size = 1) +
  scale_colour_viridis(option = "plasma")
```

## Violin plot mutation rate
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
Idents(s.obj) <- "seurat_clusters"
VlnPlot(object = with.mutation.rate.s.obj, features = c("mutation_rate"))
```

## Distribution of mutation rates in all samples
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
sc.clonedf %>% ggplot(aes(x = count, y = mutation_rate)) + geom_point()
```

## Clone size (>= 10) vs number of mutations
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
sc.clonedf %>% subset(count >= 10) %>% ggplot(aes(x = count, y = mutation_rate)) + geom_point()
```

```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
# select clones that have more than 10 cells only.
selected.clones <- subset(sc.clonedf, sc.clonedf$count >= 10)$Clone

for (clone.id in selected.clones){
  selected.cells <- subset(s.obj@meta.data, s.obj@meta.data[[sprintf("VJcombi_CDR3_%s", thres)]] == clone.id) %>% rownames()
  highligh.list <- list(group1 = selected.cells,
                        others = setdiff(row.names(s.obj@meta.data), selected.cells))
  
  names(highligh.list) <- c(clone.id, "others")
  p <- DimPlot(object = s.obj,
          reduction = "INTE_UMAP",
          label = TRUE,
          label.box = TRUE,
          repel = TRUE,
          cols.highlight = c("red", "gray"),
          cells.highlight = highligh.list) + ggtitle(clone.id)
  
  ggsave(plot = p, filename = sprintf("UMAP_clone_%s.svg", clone.id), device = "svg", width = 14, height = 10, dpi = 300, path = path.to.clone.distribution)
}
```

# Differential gene expression analysis between low vs. high mutation rate clones/cells
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
mutation.rate.thres <- quantile(with.mutation.rate.s.obj$mutation_rate)[["75%"]]
high.mutation.rate.cells <- subset(with.mutation.rate.s.obj@meta.data, with.mutation.rate.s.obj@meta.data$mutation_rate >= mutation.rate.thres) %>% row.names()
low.mutation.rate.cells <- subset(with.mutation.rate.s.obj@meta.data, with.mutation.rate.s.obj@meta.data$mutation_rate < mutation.rate.thres) %>% row.names()

meta.data <- with.mutation.rate.s.obj@meta.data %>% rownames_to_column("barcode") %>%
  rowwise() %>%
  mutate(mutation.rate.group = ifelse(barcode %in% high.mutation.rate.cells, "high_mutation_rate", "low_mutation_rate")) %>%
  column_to_rownames("barcode")
meta.data <- meta.data[row.names(with.mutation.rate.s.obj@meta.data),]
with.mutation.rate.s.obj <- AddMetaData(object = with.mutation.rate.s.obj, col.name = "mutation.rate.group", metadata = meta.data$mutation.rate.group)

diff.genes.mutation.rate <- FindMarkers(object = with.mutation.rate.s.obj, ident.1 = "high_mutation_rate", ident.2 = "low_mutation_rate", group.by = "mutation.rate.group", assay = "RNA", test.use = "wilcox")
```
## UMAP: high mutation rate cells
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
DimPlot(object = with.mutation.rate.s.obj, cells.highlight = high.mutation.rate.cells) + 
  ggtitle("High mutation rate cells") + 
  theme(legend.position = "none")
```


```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
DimPlot(object = with.mutation.rate.s.obj, cells.highlight = low.mutation.rate.cells) +
  ggtitle("Low mutation rate cells") +
  theme(legend.position = "none")
```

## DGE table

```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE, fig.width=14, fig.height=10}
```